pkg/controllers/jobset_controller.go

```diff
@@ -452,30 +446,44 @@ func (r *JobSetReconciler) resumeJob(ctx context.Context, job *batchv1.Job, repl
 	}
 
 	// Get name of parent replicated job and use it to look up the pod template.
-	replicatedJobName := job.Labels[jobset.ReplicatedJobNameKey]
-	replicatedJobPodTemplate := replicatedJobTemplateMap[replicatedJobName]
-	if job.Labels != nil && job.Labels[jobset.ReplicatedJobNameKey] != "" {
+	if job.Labels != nil && job.Labels[jobset.ReplicatedJobNameKey] != "" && job.Labels[jobset.JobIndexKey] != "" {
+		jobIdx, err := strconv.Atoi(job.Labels[jobset.JobIndexKey])
+		if err != nil {
+			return err
+		}
 		// Certain fields on the Job pod template may be mutated while a JobSet is suspended,
 		// for integration with Kueue. Ensure these updates are propagated to the child Jobs
 		// when the JobSet is resumed.
-		// Merge values rather than overwriting them, since a different controller
-		// (e.g., the Job controller) may have added labels/annotations/etc to the
-		// Job that do not exist in the ReplicatedJob pod template.
+		// We merge the PodTemplate properties from the latest PodTemplate
+		// in the JobSet, with the PodTemplate corresponding to a newly created
+		// Job.
+		newJob, err := constructJob(js, rjob, jobIdx)
+		if err != nil {
+			return err
+		}
+		// Copy the Job labels managed by the Job controller
+		for _, labelKey := range jobset.JobManagedLabels {
+			newJob.Spec.Template.Labels[labelKey] = job.Spec.Template.Labels[labelKey]
+		}
 		job.Spec.Template.Labels = collections.MergeMaps(
-			job.Spec.Template.Labels,
-			replicatedJobPodTemplate.Labels,
+			newJob.Spec.Template.Labels,
+			rjob.Template.Spec.Template.Labels,
```

Comment:
What if we put job.Spec.Template.Labels[labelKey] in the MergeMaps?

Reply:
I just don't understand why we should selectively only copy the JobManagedLabels.

Wouldn't that mean that we dropping the original job labels and only preserving the JobManagedLabels?

If I was building a custom CRD using JobSet and I had custom labels and they were dropped, that would be a bug to me.

Reply:
> I just don't understand why we should selectively only copy the JobManagedLabels.

I feel this is a very similar question as already being discussed above. Take a look at the most recent answer there: https://github.com/kubernetes-sigs/jobset/pull/640/files/4b9625725820a015d29bbcb65b114a5baa032d29#r1711840934. Basically, I merge labels from two sources: JobSet template, and oldJob (but only limited to `JobManagedLabels`, because those are added by API server.